libndt7/src/main/java/net/measurementlab/ndt7/android/Uploader.kt

```diff
@@ -0,0 +1,116 @@
+@file:JvmName("Uploader")
+package net.measurementlab.ndt7.android
+
+import com.google.gson.Gson
+import net.measurementlab.ndt7.android.utils.DataConverter.currentTimeInMicroseconds
+import net.measurementlab.ndt7.android.utils.DataConverter.generateResponse
+import net.measurementlab.ndt7.android.utils.NDT7Constants.MAX_RUN_TIME
+import net.measurementlab.ndt7.android.utils.NDT7Constants.MEASUREMENT_INTERVAL
+import net.measurementlab.ndt7.android.utils.NDT7Constants.MIN_MESSAGE_SIZE
+import net.measurementlab.ndt7.android.NDTTest.TestType.*
+import net.measurementlab.ndt7.android.models.CallbackRegistry
+import net.measurementlab.ndt7.android.models.Measurement
+import net.measurementlab.ndt7.android.utils.PayloadTransformer
+import net.measurementlab.ndt7.android.utils.SocketFactory
+import okhttp3.OkHttpClient
+import okhttp3.WebSocket
+import okhttp3.WebSocketListener
+import okio.ByteString
+import java.net.URI
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Semaphore
+
+class Uploader(
+        private val cbRegistry: CallbackRegistry,
+        private val executorService: ExecutorService,
+        private val speedtestLock: Semaphore
+) : WebSocketListener() {
+
+    private var startTime: Long = 0
+    private var previous: Long = 0
+    private var totalBytesSent = 0.0
+    private val gson = Gson()
+
+    override fun onMessage(webSocket: WebSocket, text: String) {
+
+        try {
+            val measurement = gson.fromJson(text, Measurement::class.java)
+            cbRegistry.measurementProgressCbk(measurement)
+        } catch (e: Exception) {
+            //we don't care that much if a single measurement has trouble
+            return
+        }
+    }
+
+
+override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
+
+    val clientResponse = generateResponse(startTime, totalBytesSent, UPLOAD)
+    when (code) {
+        1000 -> {
+            cbRegistry.onFinishedCbk(clientResponse, null, UPLOAD)
+        }
+        else -> {
+            cbRegistry.onFinishedCbk(clientResponse, Error(reason), UPLOAD)
+        }
+    }
+
+    releaseResources()
+    webSocket.close(1000, null)
+}
```

Comment:
```suggestion
    override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {

        val clientResponse = generateResponse(startTime, totalBytesSent, UPLOAD)
        when (code) {
            1000 -> {
                cbRegistry.onFinishedCbk(clientResponse, null, UPLOAD)
            }
            else -> {
                cbRegistry.onFinishedCbk(clientResponse, Error(reason), UPLOAD)
            }
        }

        releaseResources()
        webSocket.close(1000, null)
    }
```

(it seems this function needs formatting?)

Reply:
Yeah thats frustrating. We really need to add a linter so this isn't a manual process. I might do that real quickly right now. I should've done that before opening up the PR in the first place